File: algos/cloneGraph/cloneGraph.py

class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


# return a deep clone of graph.

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node: return None
        
        oldToNew = {}
        # could implement as stack or queue, doesn;t really matter
        # implements like { oldNode: newNode }
        stack = [node]

        while stack:
            cur = stack.pop()
            if cur not in oldToNew:
                clone = Node(cur.val)
                oldToNew[cur] = clone
            else:
                clone = oldToNew[cur]
            
            for neighbor in cur.neighbors:
                if neighbor not in oldToNew:
                    stack.append(neighbor)
                    newNeighbor = Node(neighbor.val)
                    oldToNew[neighbor] = newNeighbor
                clone.neighbors.append(oldToNew[neighbor])
        
        return oldToNew[node]
        

File: algos/cloneGraph/cloneGraph.ts
class Noode {
  val: number;
  neighbors: Noode[];
  constructor(val?: number, neighbors?: Noode[]) {
    this.val = val === undefined ? 0 : val;
    this.neighbors = neighbors === undefined ? [] : neighbors;
  }
}

type OldToNew = Map<Noode, Noode>;

function cloneGraph(node: Noode | null): Noode | null {
  if (!node) return null;

  const oldToNew: OldToNew = new Map();
  const stack = [node];

  while (stack.length) {
    const cur = stack.pop();
    if (!cur) continue;
    let clone: Noode;

    if (!oldToNew.has(cur)) {
      clone = new Noode(cur.val);
      oldToNew.set(cur, clone);
    } else {
      clone = oldToNew.get(cur)!;
    }

    for (let neighbor of cur.neighbors) {
      if (!oldToNew.has(neighbor)) {
        const newNeighbor = new Noode(neighbor.val);
        oldToNew.set(neighbor, newNeighbor);
        stack.push(newNeighbor);
      }
      clone.neighbors.push(oldToNew.get(neighbor)!);
    }
  }

  return oldToNew.get(node)!;
}
